specs/lazy-adr/adr-018-forced-inclusion-mechanism.md

```diff
@@ -0,0 +1,462 @@
+3. **Block Processing Logic**: Must implement the modified fork choice rule
+4. **DA Client**: Must be enhanced to scan for direct transactions
+5. **Transaction Validation**: Must validate both sequencer-batched and direct transactions
+
+### Data Structures
+
+#### Leveraging ExtraData in Block Header
+
+As defined in ADR-015 (Rollkit Minimal Header), the Rollkit minimal header already includes an `ExtraData` field designed for additional metadata. We will leverage this field instead of adding new explicit fields to the header structure:
+
+```go
+type Header struct {
+    // Existing fields as defined in ADR-015
+    ParentHash Hash
+    Height uint64
+    Timestamp uint64
+    ChainID string
+    DataCommitment []byte
+    StateRoot Hash
+    ExtraData []byte // Will contain DA reference information in a structured format
+}
+```
+
+For the forced inclusion mechanism, we will encode the following information in the `ExtraData` field:
+
+```go
+type ForcedInclusionExtraData struct {
+    DAOriginHeight    uint64         // Height of the DA layer block this rollup block derives from
+    DAOriginHash      []byte         // Hash of the DA layer block this rollup block derives from
+    DAOriginTimestamp uint64         // Timestamp of the DA layer block
+    DirectTxsIncluded []common.Hash  // Hashes of direct transactions included from the DA layer
+    // Other existing extraData fields, such as sequencer information
+}
+```
+
+This approach maintains compatibility with the minimal header structure while adding the necessary information for forced inclusion.
+
+#### Direct Transaction Tracking
+
+```go
+type DirectTransaction struct {
+    TxHash       common.Hash  // Hash of the transaction
+    DAHeight     uint64       // Height at which it was included in the DA layer
+    DAHash       []byte       // Hash of the DA block
+    Included     bool         // Whether it has been included in a rollup block
+    IncludedAt   uint64       // Height at which it was included in the rollup
+}
+
+type DirectTxTracker struct {
+    txs         map[common.Hash]DirectTransaction  // Map of direct transactions
+    mu          sync.RWMutex                       // Mutex for thread-safe access
+    latestDAHeight uint64                          // Latest DA block height scanned
+}
+```
+
+#### Sequencer Status Tracking
+
+```go
+type SequencerStatus struct {
+    IsActive          bool      // Whether the sequencer is considered active
+    LastActiveHeight  uint64    // Last DA height where sequencer posted a batch
+    DownSince         time.Time // Time since sequencer was last seen active
+}
+```
+
+### APIs and Interfaces
+
+#### Enhanced DA Client Interface
+
+```go
+type DAClient interface {
+    // Existing methods
+    // ...
+    
+    // New method for forced inclusion
+    GetDirectTransactions(ctx context.Context, fromHeight, toHeight uint64) ([][]byte, error)
```

Comment:
we should use nodes subscription system to subscribe to a namespace. this would probably be a simpler design overall as well.

Reply:
Yes, if there's a simple way to subscribe to a Celestia Node, we can do this. But afaik, that doesn't exist yet. Correct me if wrong @gupadhyaya @tzdybal